File: csharp/csharp/2023/Day1/Day1.cs
using System.Text.RegularExpressions;
using csharp.csharp_lib;
using FluentAssertions;

namespace csharp._2023.Day1;

static class Day1
{
    public static void Part1()
    {
        var input = Utilities.GetLines("/2023/Day1/Data.txt");
        var regex = new Regex(@"\d");

        var result = input.Select(line => regex.Matches(line))
            .Select(matches => matches.First().Value + matches.Last().Value)
            .Select(int.Parse)
            .Sum();

        result.Should().Be(55029);
    }

    public static void Part2()
    {
        var regex = new Regex(@"\d");
        var input = Utilities.GetLines("/2023/Day1/Data.txt");
        var res = input
            .Select(ReplaceWordsWithLetters);

        var res2 = res.Select(line => regex.Matches(line))
                .Select(matches => matches.First().Value + matches.Last().Value)
                .Select(int.Parse);

        var result = res2.Sum();
        result.Should().Be(55686);
        Console.WriteLine(result);
    }

    private static string ReplaceWordsWithLetters(string input)
    {
        return input
            .Replace("one", "o1e")
            .Replace("two", "t2o")
            .Replace("three", "th3e")
            .Replace("four", "4")
            .Replace("five", "5e")
            .Replace("six", "6")
            .Replace("seven", "7n")
            .Replace("eight", "e8t")
            .Replace("nine", "n9e");
    }
}

File: csharp/fsharp/2023/Day1/Day1.fs
module fsharp.Day1

open System.Text.RegularExpressions

let regex = Regex("\d")

let fixInvalidNumbers (s: string) =
    s
        .Replace("one", "o1e")
        .Replace("two", "t2o")
        .Replace("three", "th3e")
        .Replace("four", "4")
        .Replace("five", "5e")
        .Replace("six", "6")
        .Replace("seven", "7n")
        .Replace("eight", "e8t")
        .Replace("nine", "n9e")

let findNumbers (input: string) (regex: Regex) =
    let matches = regex.Matches(input)
    let firstMatch = matches.Item(0).Value
    let lastMatch = matches.Item(matches.Count - 1).Value

    int $"{firstMatch}{lastMatch}"

let input = System.IO.File.ReadAllLines "2023/Day1/Data.txt"

let part1Res = input |> Seq.map (fun x -> findNumbers x regex) |> Seq.sum
assert (part1Res = 55029)

let part2Res =
    input
    |> Seq.map fixInvalidNumbers
    |> Seq.map (fun x -> findNumbers x regex)
    |> Seq.sum

assert (part2Res = 55686)
